cluster-autoscaler/cloudprovider/clusterapi/internal/clusterapi_machineset.go

```diff
@@ -0,0 +1,150 @@
+package internal
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/golang/glog"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	v1alpha1apis "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+)
+
+const (
+	machineDeleteAnnotationKey = "sigs.k8s.io/cluster-api-delete-machine"
+)
+
+type clusterMachineSet struct {
+	*clusterManager
+	*v1alpha1apis.MachineSet
+	min     int
+	max     int
+	nodes   []string
+	nodeSet map[string]bool
+}
+
+func (m *clusterMachineSet) Name() string {
+	return m.MachineSet.Name
+}
+
+func (m *clusterMachineSet) Namespace() string {
+	return m.MachineSet.Namespace
+}
+
+func (m *clusterMachineSet) MinSize() int {
+	return m.min
+}
+
+func (m *clusterMachineSet) MaxSize() int {
+	return m.max
+}
+
+func (m *clusterMachineSet) Replicas() int {
+	if m.MachineSet.Spec.Replicas == nil {
+		return 0
+	}
+	glog.Infof("machineset: %q has %d replicas", m.MachineSet.Name, *m.MachineSet.Spec.Replicas)
+	return int(*m.MachineSet.Spec.Replicas)
+}
+
+func (m *clusterMachineSet) SetSize(nreplicas int) error {
+	ms, err := m.clusterapi.MachineSets(m.MachineSet.Namespace).Get(m.MachineSet.Name, v1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("Unable to get machineset %q: %v", m.MachineSet.Name, err)
+	}
+
+	newMachineSet := ms.DeepCopy()
+	replicas := int32(nreplicas)
+	newMachineSet.Spec.Replicas = &replicas
+
+	_, err = m.clusterapi.MachineSets(m.MachineSet.Namespace).Update(newMachineSet)
+	if err != nil {
+		return fmt.Errorf("Unable to update number of replicas of machineset %q: %v", m.MachineSet.Name, err)
+	}
+
+	return nil
+}
+
+func (m *clusterMachineSet) Nodes() ([]string, error) {
+	return m.nodes, nil
+}
+
+func (m *clusterMachineSet) DeleteNodes(nodenames []string) error {
+	if len(nodenames) == 0 {
+		return nil
+	}
+
+	snapshot := m.getClusterState()
+
+	for _, nodename := range nodenames {
+		name, exists := snapshot.NodeMap[machineSetID(m.MachineSet)][nodename]
+		if !exists {
+			return fmt.Errorf("cannot map nodename %q to machine", nodename)
+		}
+		machine, err := m.clusterapi.Machines(m.MachineSet.Namespace).Get(name, v1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("cannot get machine %s/%s: %v", m.MachineSet.Namespace, name, err)
+		}
+
+		machine = machine.DeepCopy()
+
+		if machine.Annotations == nil {
+			machine.Annotations = map[string]string{}
+		}
+
+		// Annotate machine that it is the chosen one.
+		// TODO(frobware) - don't hardcode this name
+		machine.Annotations["sigs.k8s.io/cluster-api-delete-machine"] = time.Now().String()
```

Comment:
This is going to change soon most likely. Upstream will not accept this way of deleting machines. It's more preferred to use atomic operation. E.g. just update a single object, in this case a machineset manifest as I described in https://github.com/kubernetes-sigs/cluster-api/pull/513#issuecomment-426400245.